### PR TITLE
Update docs on declaring env key dependencies

### PR DESCRIPTION
**Pull request by mehulkar**
Docs for `env` key updates

- [x] #1970
- [x] #1950
- [ ] codemod
- [ ] lint rule

This PR will include: 

- guide and reference docs updates
- blog post

This PR will be ready for review after all the code pieces are merged.

### CODE REVIEW

File: docs/pages/docs/core-concepts/caching.mdx [hunk 1]
```diff
@@ -155,49 +155,67 @@ determining a cache hit on the `test` task:
...
+
+## Altering Caching Based on Environment Variables
+
 When you use `turbo` with tools that inline environment variables at build time (e.g. [Next.js](https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser) or [Create React App](https://create-react-app.dev/docs/adding-custom-environment-variables/)), it is important to tell `turbo` about it. Otherwise, you could ship a cached artifact with the wrong environment variables!
 
-Luckily, you can control `turbo`'s cache fingerprinting (also known as "hashing") behavior based on the values of both environment variables and the contents of files:
+Luckily, you can control `turbo`'s cache fingerprinting (also known as "hashing") behavior based on the values of environment variables.
 
-- Including environment variables in a `dependsOn` in your `pipeline` definition prefixed by a `$` will impact the cache fingerprint on a per-task or per-workspace-task basis.
+- Including environment variables in a `env` in your `pipeline` definition will impact the cache fingerprint on a per-task or per-workspace-task basis.
```
Comment (nathanhammond) [inline docs/pages/docs/core-concepts/caching.mdx:183]:
```suggestion
- Including environment variables in the `env` key in your `pipeline` definition will impact the cache fingerprint on a per-task or per-workspace-task basis.
```

File: docs/pages/docs/core-concepts/caching.mdx [hunk 2]
```diff
@@ -209,6 +227,11 @@ Luckily, you can control `turbo`'s cache fingerprinting (also known as "hashing"
 }
 ```
 
+<Callout type="info">
+  It is also possible to declare environment variables in the `dependsOn` config with a `$` prefix for
```
Comment (nathanhammond) [inline docs/pages/docs/core-concepts/caching.mdx:231]: Let's not present this in a "positive" framing ("it is also possible") and instead jump right into "is deprecated."

File: docs/pages/docs/reference/configuration.mdx [hunk 3]
```diff
@@ -98,24 +103,26 @@ Prefixing an item in `dependsOn` with a `$` tells `turbo` that this pipeline tas
...
+      "env": ["$SOMETHING_ELSE"], // value will impact the hashes of all build tasks
       "outputs": ["dist/**", ".next/**"]
     },
     "web#build": {
-      "dependsOn": [
-        "^build",
-        "$STRIPE_SECRET_KEY", // value will impact hash of only web's build task
-      ],
+      "dependsOn": ["^build"],
+      "env": ["$STRIPE_SECRET_KEY"], // value will impact hash of only web's build task
```
Comment (nathanhammond) [inline docs/pages/docs/reference/configuration.mdx:125]:
```suggestion
      "env": ["STRIPE_SECRET_KEY"], // value will impact hash of only web's build task
```

File: docs/pages/docs/reference/configuration.mdx [hunk 4]
```diff
@@ -98,24 +103,26 @@ Prefixing an item in `dependsOn` with a `$` tells `turbo` that this pipeline tas
...
 {
   "$schema": "https://turborepo.org/schema.json",
   "pipeline": {
     "build": {
-      "dependsOn": [
-        "^build",
-        "$SOMETHING_ELSE" // value will impact the hashes of all build tasks
-      ],
+      "dependsOn": ["^build"],
+      "env": ["$SOMETHING_ELSE"], // value will impact the hashes of all build tasks
```
Comment (nathanhammond) [inline docs/pages/docs/reference/configuration.mdx:120]:
```suggestion
      "env": ["SOMETHING_ELSE"], // value will impact the hashes of all build tasks
```

File: docs/schema.d.ts [hunk 5]
```diff
@@ -58,13 +58,21 @@ export interface Pipeline {
    * package level (e.g. "a package's test and lint commands depend on build being
    * completed first").
    *
-   * Prefixing an item in dependsOn with a $ tells turbo that this pipeline task depends
-   * the value of that environment variable.
+   * DEPRECATED: Items prefixed with a `$` tells turbo that this pipeline task depends
+   * the value of that environment variable. This feature is deprecated and the `env` key
+   * should be used instead.
```
Comment (nathanhammond) [inline docs/schema.d.ts:63]: This is the "just-in-time" documentation, let's avoid telling them about the deprecated feature? @mattpocock to second guess me, please.
```suggestion
```

File: docs/schema.d.ts [hunk 6]
```diff
@@ -58,13 +58,21 @@ export interface Pipeline {
...
    *
    * @default []
    */
   dependsOn?: string[];
 
+  /**
+   * The list of environment variables this task depends on.
```
Comment (nathanhammond) [inline docs/schema.d.ts:70]: If we are going to require/deny using a $ prefix it should be noted here.

**Review by nathanhammond — COMMENTED** [review]
There are also a number of places in `/examples` that would need to be updated.

**Comment by mehulkar** [comment]
> There are also a number of places in /examples that would need to be updated.

Do we want to put empty arrays in `examples/`? I only found one actual example, and it's one that isn't needed anymore after [automatic inclusion](https://turborepo.org/blog/turbo-1-4-0#automatic-environment-variable-inclusion)

**Review by tknickman — APPROVED** [review]

**Pull request merged**